### PR TITLE
Translate complex nested vector expressions instead of lowering them

### PR DESCRIPTION
**Pull request by vmaksimo**
Allow the translator to go deeper into complex nested instructions.
Enabled a possibility to translate expressions in `extractelement` and binary operator instructions (e.g., `fadd`, `fmul`).
Removed lowering of constant expression vectors as these two methods are mutually exclusive.

This patch was developed as an alternative to this fix: #1330

### CODE REVIEW

**Comment by vmaksimo** [comment]
Tagging @AlexeySotkin @MrSidims @pauzinl @KornevNikita to take a look.

File: lib/SPIRV/libSPIRV/SPIRVInstruction.h [hunk 1]
```diff
@@ -2757,11 +2757,28 @@ _SPIRV_OP(ImageQuerySamples, true, 4)
...
+  std::vector<SPIRVEntry *> getNonLiteralOperands() const override {
+    // If SpecConstant results from CompositeExtract/Insert operation, then all
+    // operands are expected to be literals.
+    SPIRVWord LiteralOpcode = Ops[0];
+    if (LiteralOpcode == OpCompositeExtract ||
+        LiteralOpcode == OpCompositeInsert)
+      return {};
+
+    return SPIRVInstTemplateBase::getNonLiteralOperands();
+  }
```
Comment (AlexeySotkin) [inline lib/SPIRV/libSPIRV/SPIRVInstruction.h:2775]: Still, we may call `isOperandLiteral()` which will return wrong answer. I suggest overriding `isOperandLiteral()` instead
```suggestion
  bool isOperandLiteral(unsigned I) const override {
    switch (Ops[0]) { // Opcode
    case OpCompositeExtract:
    case OpCompositeInsert:
      return true;
    default:
      return SPIRVInstTemplateBase::isOperandLiteral(I); // Will it always return false?
    }
  }
```

Comment (vmaksimo) [reply lib/SPIRV/libSPIRV/SPIRVInstruction.h:2775]: Nice suggestion, thanks! Not sure if `SPIRVInstTemplateBase::isOperandLiteral(I)` will always return `false`, but it looks right to call the base class method to be safe for sure.

**Comment by AlexeySotkin** [comment]
@vmaksimo could you please create a PR with this patch at intel/llvm repo and run tests there to make sure this patch will not bring any regressions when we integrate it to the SYCL compiler?

**Review by AlexeySotkin — APPROVED** [review]

**Comment by AlexeySotkin** [comment]
@zuban32 FYI.

**Pull request merged**